rust/connlib/tunnel/src/client.rs

```diff
@@ -576,43 +592,54 @@ impl ClientState {
 
         self.gateways_site.insert(gateway_id, site_id);
         self.recently_connected_gateways.put(gateway_id, ());
+        self.peers
+            .add_ips_with_resource(&gateway_id, &ips, &resource_id);
 
-        if self.peers.get(&gateway_id).is_some() {
-            self.peers
-                .add_ips_with_resource(&gateway_id, &ips, &resource_id);
-
-            self.buffered_events.push_back(ClientEvent::RequestAccess {
-                resource_id,
+        if !self.node.has_connection(gateway_id) {
+            let (offer, buffer) = self.node.new_connection(
                 gateway_id,
-                maybe_domain: awaiting_connection_details.domain,
-            });
-            return Ok(());
-        };
+                awaiting_connection_details.last_intent_sent_at,
+                now,
+            );
+            buffer.extend(awaiting_connection_details.packets);
 
-        self.peers.insert(
-            GatewayOnClient::new(gateway_id, &ips, HashSet::from([resource_id])),
-            &[],
-        );
-        self.peers
-            .add_ips_with_resource(&gateway_id, &ips, &resource_id);
+            self.buffered_events
+                .push_back(ClientEvent::RequestConnection {
+                    resource_id,
+                    gateway_id,
+                    preshared_key: Secret::new(Key(*offer.session_key.expose_secret())),
+                    offer: Offer {
+                        username: offer.credentials.username,
+                        password: offer.credentials.password,
+                    },
+                    maybe_domain: awaiting_connection_details.domain,
+                });
```

Comment:
In the future, we can split this here into a dedicated `RequestConnection` message and a parallel `RequestAccess` (once #6403 is widely deployed). That will allow us to remove the duplicated resource stuff from `RequestConnection`.